### PR TITLE
EP-560 Daemon can get bluzelle public key from Solidity contract on E…

### PR DESCRIPTION
**Pull request by rnistuk**
…thereum

Also snuck in the fix for the protobuf issue on macOS.

### CODE REVIEW

File: utils/crypto.cpp [hunk 1]
```diff
@@ -178,6 +183,111 @@ namespace
             return false;
         }
     }
+
+
+    size_t
+    write_function(void* contents,size_t size, size_t nmemb, void* userp)
```
Comment (ebruck) [inline utils/crypto.cpp]: nmemb?

File: utils/crypto.cpp [hunk 2]
```diff
@@ -178,6 +183,111 @@ namespace
...
+    {
+        reinterpret_cast<std::string*>(userp)->append(reinterpret_cast<char*>(contents), size * nmemb);
+        return size * nmemb;
+    }
+
+
+    bzn::json_message
+    get_curl_response(const std::string& function_call, const std::string& url = ROPSTEN_URL)
+    {
+        const std::string MSG_ERROR_CURL{"curl_easy_perform() failed: "};
```
Comment (ebruck) [inline utils/crypto.cpp]: I would put this outside so it's initialized only once.

File: utils/crypto.cpp [hunk 3]
```diff
@@ -178,6 +183,111 @@ namespace
...
+        }
+        return ss.str();
+    }
+
+
+    // TODO rename hex_string_to_string
+    std::string
+    hex_string_to_string(const std::string& hex_string)
+    {
+        std::string public_key{""};
```
Comment (ebruck) [inline utils/crypto.cpp]: init not required

File: utils/crypto.cpp [hunk 4]
```diff
@@ -186,8 +296,20 @@ namespace bzn::utils::crypto
     std::string
     retrieve_bluzelle_public_key_from_contract()
     {
+        std::string public_key_hex{""};
         // TODO retrieve the public key from the contract
-        return temporary_public_pem;
+        const size_t key_size = get_public_pem_size();
```
Comment (ebruck) [inline utils/crypto.cpp]: why the temp? You don't it elsewhere.

File: utils/crypto.cpp [hunk 5]
```diff
@@ -186,8 +296,20 @@ namespace bzn::utils::crypto
     std::string
     retrieve_bluzelle_public_key_from_contract()
     {
+        std::string public_key_hex{""};
```
Comment (ebruck) [inline utils/crypto.cpp]: unnecessary init

File: utils/crypto.cpp [hunk 6]
```diff
@@ -178,6 +183,111 @@ namespace
...
+        curl_easy_setopt(curl.get(), CURLOPT_POSTFIELDS, post_fields.c_str());
+
+        curl_easy_setopt(curl.get(), CURLOPT_POSTFIELDSIZE, -1L);
+        curl_easy_setopt(curl.get(), CURLOPT_URL, url.c_str());
+        curl_easy_setopt(curl.get(), CURLOPT_WRITEFUNCTION, reinterpret_cast<void*>(write_function));
+
+        CURLcode res = curl_easy_perform(curl.get());
+        if (res != CURLE_OK)
+        {
+            throw (std::runtime_error(MSG_ERROR_CURL + curl_easy_strerror(res)));
```
Comment (ebruck) [inline utils/crypto.cpp]: Is this truly a fatal error?

File: utils/crypto.cpp [hunk 7]
```diff
@@ -178,6 +183,111 @@ namespace
...
+        }
+
+        bzn::json_message response;
+        Json::Reader reader;
+
+        if (!reader.parse(readBuffer, response))
+        {
+            LOG(error) << "Unable to parse response from Ropsten - could not get key size";
```
Comment (ebruck) [inline utils/crypto.cpp]: If the above is a fatal error then why don't we throw here?

File: utils/crypto.cpp (hunk 2 above)
Comment (rnistuk) [reply utils/crypto.cpp]: OK.  Done.

File: utils/crypto.cpp (hunk 3 above)
Comment (rnistuk) [reply utils/crypto.cpp]: Removed.

File: utils/crypto.cpp (hunk 4 above)
Comment (rnistuk) [reply utils/crypto.cpp]: Thanks, good catch. Removed.

File: utils/crypto.cpp (hunk 5 above)
Comment (rnistuk) [reply utils/crypto.cpp]: Removed.

File: utils/crypto.cpp (hunk 6 above)
Comment (rnistuk) [reply utils/crypto.cpp]: Probably not. I think I can just do what I do in the net block. I'll move response up a bit...

File: utils/crypto.cpp (hunk 7 above)
Comment (rnistuk) [reply utils/crypto.cpp]: Changed the previous throw to a LOG(error)

**Pull request merged**